<robot name="spot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <link name="base_link">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/base.obj"/>
      </geometry>
      <material name="base_link">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/base.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="16.7"/>
        <inertia ixx="0.081" ixy="0.0" ixz="-0.001" iyy="0.646" iyz="0.0" izz="0.666"/>
    </inertial>
    <dynamics damping="10" friction="0.8"/>
  </link>

  <!-- Box Control Plugin-->
  <gazebo>
    <plugin name="spot_box_control" filename="libspot_box_control.so">
    </plugin>
  </gazebo>

  <!-- Camera Definitions -->
  <gazebo reference="base_link">
    <sensor name="right_camera" type="camera">
      <pose>-0.125 -0.12 0.035 0 0.2 -1.571</pose>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>800</width>
          <height>600</height>
        </image>
      </camera>
      <always_on>1</always_on>
      <update_rate>10</update_rate>
      <visualize>true</visualize>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>spot</robotNamespace>
        <cameraName>right_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="left_camera" type="camera">
      <pose>-0.125 0.12 0.035 0 0.2 1.571</pose>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>800</width>
          <height>600</height>
        </image>
      </camera>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>spot</robotNamespace>
        <cameraName>left_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="front_right_camera" type="camera">
      <pose>0.415 -0.037 -0.01 -1.355558 0.353917 0.554196</pose>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>800</width>
          <height>600</height>
        </image>
      </camera>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>spot</robotNamespace>
        <cameraName>front_right_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="back_camera" type="camera">
      <pose>-0.425 0 0.01 0 0.3 -3.142</pose>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>800</width>
          <height>600</height>
        </image>
      </camera>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>spot</robotNamespace>
        <cameraName>back_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="front_left_camera" type="camera">
      <pose>0.415 0.037 -0.01 1.355558 0.353917 -0.554196</pose>
      <camera>
        <horizontal_fov>1.047</horizontal_fov>
        <image>
          <width>800</width>
          <height>600</height>
        </image>
      </camera>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>true</visualize>
      <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
        <robotNamespace>spot</robotNamespace>
        <cameraName>front_left_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
      </plugin>
    </sensor>
  </gazebo>

  <!-- Front Left Hip -->
  <link name="front_left.hip">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fl.hip.obj"/>
      </geometry>
      <material name="front_left.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fl.hip.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="2.5"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_left.hx" type="fixed">
    <origin xyz="0.29785 0.05500 0.00000" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="front_left.hip"/>
  </joint>

  <!-- Front Left Upper Leg -->
  <link name="front_left.upperleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fl.uleg.obj"/>
      </geometry>
      <material name="front_left.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fl.uleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="1.8"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_left.hy" type="fixed">
    <origin xyz="0.0 0.11080 0.00" rpy="0 0 0"/>
    <parent link="front_left.hip"/>
    <child link="front_left.upperleg"/>
  </joint>

  <!-- Front Left Lower Leg -->
  <link name="front_left.lowerleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.lleg.obj"/>
      </geometry>
      <material name="front_left.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.lleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="0.8"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_left.knee" type="fixed">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <parent link="front_left.upperleg"/>
    <child link="front_left.lowerleg"/>
  </joint>

  <!-- Front Right Hip -->
  <link name="front_right.hip">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.hip.obj"/>
      </geometry>
      <material name="front_right.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.hip.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="2.5"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_right.hx" type="fixed">
    <origin xyz="0.29785 -0.05500 0.00000" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="front_right.hip"/>
  </joint>

  <!-- Front Right Upper Leg -->
  <link name="front_right.upperleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.uleg.obj"/>
      </geometry>
      <material name="front_right.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.uleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="1.8"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_right.hy" type="fixed">
    <origin xyz="0.0 -0.11080 0.00" rpy="0 0 0"/>
    <parent link="front_right.hip"/>
    <child link="front_right.upperleg"/>
  </joint>

  <!-- Front Right Lower Leg -->
  <link name="front_right.lowerleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.lleg.obj"/>
      </geometry>
      <material name="front_right.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/fr.lleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="0.8"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="front_right.knee" type="fixed">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <parent link="front_right.upperleg"/>
    <child link="front_right.lowerleg"/>
  </joint>

  <!-- Hind Left Hip -->
  <link name="hind_left.hip">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.hip.obj"/>
      </geometry>
      <material name="hind_left.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.hip.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="2.5"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_left.hx" type="fixed">
    <origin xyz="-0.29785 0.05500 0.00" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="hind_left.hip"/>
  </joint>

  <!-- Hind Left Upper Leg -->
  <link name="hind_left.upperleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.uleg.obj"/>
      </geometry>
      <material name="hind_left.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.uleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="1.8"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_left.hy" type="fixed">
    <origin xyz="0.0 0.11080 0.00" rpy="0 0 0"/>
    <parent link="hind_left.hip"/>
    <child link="hind_left.upperleg"/>
  </joint>

  <!-- Hind Left Lower Leg -->
  <link name="hind_left.lowerleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.lleg.obj"/>
      </geometry>
      <material name="hind_left.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hl.lleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="0.8"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_left.knee" type="fixed">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <parent link="hind_left.upperleg"/>
    <child link="hind_left.lowerleg"/>
  </joint>

  <!-- Hind Right Hip -->
  <link name="hind_right.hip">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.hip.obj"/>
      </geometry>
      <material name="hind_right.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.hip.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="2.5"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_right.hx" type="fixed">
    <origin xyz="-0.29785 -0.05500 0.00000" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="hind_right.hip"/>
  </joint>

  <!-- Hind Right Upper Leg -->
  <link name="hind_right.upperleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.uleg.obj"/>
      </geometry>
      <material name="hind_right.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.uleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="1.8"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_right.hy" type="fixed">
    <origin xyz="0.0 -0.11080 0.00" rpy="0 0 0"/>
    <parent link="hind_right.hip"/>
    <child link="hind_right.upperleg"/>
  </joint>

  <!-- Hind Right Lower Leg -->
  <link name="hind_right.lowerleg">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.lleg.obj"/>
      </geometry>
      <material name="hind_right.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_IL_environment/models/spot/meshes/hr.lleg.obj"/>
      </geometry>
    </collision>
    <inertial>
        <mass value="0.8"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
    <dynamics damping="10" friction="0.5"/>
  </link>

  <joint name="hind_right.knee" type="fixed">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <parent link="hind_right.upperleg"/>
    <child link="hind_right.lowerleg"/>
  </joint>

</robot>
